flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/catalog/ManagedTableListener.java

```diff
@@ -0,0 +1,125 @@
+import org.apache.flink.table.factories.DynamicTableFactory;
+import org.apache.flink.table.factories.FactoryUtil;
+import org.apache.flink.util.StringUtils;
+
+import javax.annotation.Nullable;
+
+import java.util.Map;
+
+import static org.apache.flink.table.catalog.CatalogBaseTable.TableKind.MANAGED;
+import static org.apache.flink.table.factories.ManagedTableFactory.discoverManagedTableFactory;
+
+/** The listener for managed table operations. */
+public class ManagedTableListener {
+
+    private final ClassLoader classLoader;
+
+    private final ReadableConfig config;
+
+    public ManagedTableListener(ClassLoader classLoader, ReadableConfig config) {
+        this.classLoader = classLoader;
+        this.config = config;
+    }
+
+    /** Notify for creating managed table. */
+    public ResolvedCatalogBaseTable<?> notifyTableCreation(
+            @Nullable Catalog catalog,
+            ObjectIdentifier identifier,
+            ResolvedCatalogBaseTable<?> table,
+            boolean isTemporary,
+            boolean ignoreIfExists) {
+        if (isNewlyManagedTable(catalog, table)) {
+            ResolvedCatalogTable managedTable = createManagedTable(identifier, table, isTemporary);
+            discoverManagedTableFactory(classLoader)
+                    .onCreateTable(
+                            createTableFactoryContext(identifier, managedTable, isTemporary),
+                            ignoreIfExists);
+            return managedTable;
+        }
+        return table;
+    }
+
+    /** Notify for dropping managed table. */
+    public void notifyTableDrop(
+            ObjectIdentifier identifier,
+            ResolvedCatalogBaseTable<?> table,
+            boolean isTemporary,
+            boolean ignoreIfNotExists) {
+        if (table.getTableKind() == MANAGED) {
+            discoverManagedTableFactory(classLoader)
+                    .onDropTable(
+                            createTableFactoryContext(
+                                    identifier, (ResolvedCatalogTable) table, isTemporary),
+                            ignoreIfNotExists);
+        }
+    }
+
+    private boolean isNewlyManagedTable(
+            @Nullable Catalog catalog, ResolvedCatalogBaseTable<?> table) {
+        if (catalog == null || !catalog.supportsManagedTable()) {
+            // catalog not support managed table
+            return false;
+        }
+        if (table.getTableKind() == CatalogBaseTable.TableKind.VIEW) {
+            // view is not managed table
+            return false;
+        }
+        if (!StringUtils.isNullOrWhitespaceOnly(
+                table.getOptions().get(FactoryUtil.CONNECTOR.key()))) {
+            // with connector is not managed table
+            return false;
+        }
+        CatalogBaseTable origin = table.getOrigin();
+        // only DefaultCatalogTable or CatalogTableImpl is managed table
+        return origin instanceof DefaultCatalogTable || origin instanceof CatalogTableImpl;
```

Comment:
What's the motivation for this restriction? How does a catalog implementor learn about this?

Reply:
You are right, we should exclude `InlineCatalogTable` and `ConnectorCatalogTable` these two troublemakers.

Reply:
For `InlineCatalogTable`, it is a little tricky like:
```
try {
            origin.getOptions();
        } catch (TableException ignore) {
            // exclude abnormal tables, such as InlineCatalogTable that does not have the options
            return false;
        }
```